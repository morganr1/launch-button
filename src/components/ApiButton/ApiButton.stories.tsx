import React from 'react';
import { Story, Meta } from '@storybook/react';

import ApiButton from './ApiButton';
import { ApiButtonProps } from './ApiButton.types';

export default {
    title: 'Wayflyer/ApiButton',
    component: ApiButton,
    argTypes: {},
} as Meta<typeof ApiButton>;

const Template: Story<ApiButtonProps> = (args) => <ApiButton {...args} />;

export const Default = Template.bind({});

Default.args = {
    disabled: false,
    buttonText: {
        default: 'Launch Rocket',
        error: 'Launch Rocket',
        fetching: 'Launching',
    },
    hasError: false,
    isFetching: false,
    tooltipText: {
        default: 'Ignites the fuel',
        error: 'Ignition error',
        fetching: 'Cancel launch',
    },
    requestCallback: (data: object) => console.log('Status callback!', data),
    url: 'https://httpbin.org/delay/3',
};

export const Timeout = Template.bind({});

Timeout.args = {
    requestTimeout: 3000,
    disabled: false,
    buttonText: {
        default: 'Launch Rocket',
        error: 'Launch Rocket',
        fetching: 'Launching',
    },
    tooltipText: {
        default: 'Ignites the fuel',
        error: 'Ignition error',
        fetching: 'Cancel launch',
    },
    requestCallback: (data: any) => console.log('Status callback!', data),
    url: 'https://httpbin.org/delay/5',
};

export const Fetching = Template.bind({});

Fetching.args = {
    requestTimeout: 3000,
    disabled: false,
    isFetching: true,
    hasError: false,
    buttonText: {
        default: 'Launch Rocket',
        error: 'Launch Rocket',
        fetching: 'Launching',
    },
    tooltipText: {
        default: 'Ignites the fuel',
        error: 'Ignition error',
        fetching: 'Cancel launch',
    },
    requestCallback: (data: object) => console.log('Status callback!', data),
    url: 'https://httpbin.org/delay/5',
};

export const Error = Template.bind({});

Error.args = {
    requestTimeout: 3000,
    disabled: false,
    isFetching: false,
    hasError: true,
    buttonText: {
        default: 'Launch Rocket',
        error: 'Launch Rocket',
        fetching: 'Launching',
    },
    tooltipText: {
        default: 'Ignites the fuel',
        error: 'Ignition error',
        fetching: 'Cancel launch',
    },
    requestCallback: (data: object) => console.log('Status callback!', data),
    url: 'https://httpbin.org/delay/5',
};
