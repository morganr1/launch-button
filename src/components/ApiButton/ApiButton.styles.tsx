import styled, { keyframes } from 'styled-components';
import {
    ButtonProps,
    ButtonWrapperProps,
    StatusProps,
} from './ApiButton.types';

/* Util Functions */
export const getStatusStyles = (status: string) =>
    status === 'error' ? '#FF0000' : status === 'fetching' ? '#FF7900' : '#000';

/* Keyframes */
const Spin = keyframes`
    from {
        transform: rotate(0deg);
    }
    to {
        transform: rotate(360deg);
    }
`;

/* Styled Components */
export const Button = styled.button<ButtonProps>`
    border: 2;
    border-color: ${({ status }) => getStatusStyles(status)};
    display: inline-block;
    border-style: solid;
    background-color: #ffffff;
    width: 133px;
    height: 40px;
    cursor: pointer;
    color: ${({ status }) => getStatusStyles(status)};
    &:disabled {
        opacity: 1;
        border: 2px solid #000000;
        cursor: not-allowed;
        background-color: #f5f5f5;
        color: rgba(0, 0, 0, 0.7);
    }
`;

export const Tooltip = styled.div<StatusProps>`
    color: #ffffff;
    width: 133px;
    height: 30px;
    margin-top: 7px;
    display: flex;
    justify-content: center;
    align-items: center;
    background-color: ${({ status }) => getStatusStyles(status)};
    cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')};
    &:after {
        content: ' ';
        position: absolute;
        top: 52px;
        cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')};
        border-color: ${({ status }) => getStatusStyles(status)};
        border-width: 16px;
        border-top-width: 0px;
        border-style: solid;
        border-color: transparent transparent
            ${({ status }) => getStatusStyles(status)} transparent;
    }
`;

export const ToolTipText = styled.p`
    font-family: Roboto, sans-serif;
    font-style: normal;
    font-weight: 700;
    font-size: 12px;
    text-align: center;
    line-height: 14px;
`;

export const ButtonWrapper = styled.div<ButtonWrapperProps>`
    width: fit-content;
    height: fit-content;
`;

export const Loader = styled.div<StatusProps>`
    left: 50%;
    top: 50%;
    z-index: 1;
    width: 20px;
    margin-left: -4px;
    height: 20px;
    box-sizing: border-box;
    border: 1px solid transparent;
    border-radius: 50%;
    border-color: ${({ disabled }) => (disabled ? '#fff' : '#ff7900')}
    border-top: 1px solid;
    border-right: 1px solid;
    border-bottom: 1px solid;
    border-left: 1px solid;
    -webkit-animation: ${Spin} 1s linear infinite;
    animation: ${Spin} 1s linear infinite;
`;

export const ButtonContent = styled.div`
    display: flex;
    justify-content: space-evenly;
    align-items: center;
`;

export const ButtonText = styled.p<StatusProps>`
    margin: 0;
    padding: 0;
    font-family: Roboto, sans-serif;
    line-height: 14.06px;
    font-size: 12px;
    font-weight: 400;
    cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')}};
`;
