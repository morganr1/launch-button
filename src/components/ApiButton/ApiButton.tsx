import React, { FC, useRef, useState } from 'react';
import { ApiButtonProps, RequestStatus, HandleHover } from './AppButton.types';
import {
    Button,
    Tooltip,
    ToolTipText,
    ButtonWrapper,
    ButtonContent,
    ButtonText,
    Loader,
} from './AppButton.styles';
import { controlledFetchRequest } from './AppButton.utils';

const ApiButton: FC<ApiButtonProps> = ({
    disabled = false,
    url,
    ariaLabel,
    requestTimeout,
    isFetching = false,
    hasError = false,
    callback,
    buttonText,
    tooltipText,
    ...props
}) => {
    // It should be possible to put the button into each state via props (initial)
    const [requestStatus, setRequestStatus] = useState<RequestStatus>({
        fetching: isFetching,
        error: hasError,
    });
    const [showTooltip, setShowTooltip] = useState<boolean>(false);
    const { fetching, error } = requestStatus;
    const controllerRef = useRef<AbortController | null>();

    const handleHover: HandleHover = (show) => {
        setShowTooltip(show);
    };

    const handleClick = () => {
        if (fetching) {
            // if prop `isFetching` is initially passed as `true` - change status to error
            setRequestStatus({ error: true, fetching: false });
        }
        if (controllerRef.current) {
            // A second click of the button should abort a request that is in-flight and show the error state
            controllerRef.current.abort();
        }
        if (error || !fetching) {
            controlledFetchRequest({
                controllerRef,
                requestTimeout,
                url,
                callback,
                setRequestStatus,
            });
        }
    };
    const currentStatus = error ? 'error' : fetching ? 'fetching' : 'default';
    const allowToolTip = (error && !disabled) || (!disabled && showTooltip);

    return (
        <ButtonWrapper
            onMouseOver={() => handleHover(true)}
            onMouseOut={() => handleHover(false)}
            onFocus={() => handleHover(true)}
            onBlur={() => handleHover(false)}
            onKeyDown={({ key }) => key === 'Escape' && handleHover(false)}
        >
            <Button
                id="api-button-container"
                type="button"
                status={currentStatus}
                onClick={handleClick}
                disabled={disabled}
                aria-relevant="text"
                aria-describedby="api-button-tooltip"
                {...props}
            >
                <ButtonContent>
                    <ButtonText disabled={disabled} status={currentStatus}>
                        {buttonText[currentStatus]}
                    </ButtonText>
                    {fetching && (
                        <Loader
                            data-testid="api-button-loader"
                            status={currentStatus}
                            disabled={disabled}
                        />
                    )}
                </ButtonContent>
            </Button>
            {allowToolTip && (
                <Tooltip
                    disabled={disabled}
                    id="api-button-tooltip"
                    aria-live="polite"
                    aria-relevant="text"
                    onFocus={() => !disabled && handleHover(true)}
                    onBlur={() => !disabled && handleHover(false)}
                    onClick={handleClick}
                    status={currentStatus}
                >
                    <ToolTipText>{tooltipText[currentStatus]}</ToolTipText>
                </Tooltip>
            )}
        </ButtonWrapper>
    );
};

export default ApiButton;
