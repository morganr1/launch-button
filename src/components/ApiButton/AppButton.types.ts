/**
 * Types
 */

export type HandleHover = (show: boolean) => void;

/**
 * Interfaces
 */

export interface ControlledRequestParams {
    controllerRef: {
        current: AbortController | null | undefined;
    };
    url: string;
    setRequestStatus: Function;
    callback: Function;
    requestTimeout: number | undefined;
}

export interface RequestStatus {
    fetching: boolean;
    error: boolean;
}

export interface ApiButtonProps {
    /**
     * Text to display for each possible status
     */
    buttonText: {
        default: string;
        error: string;
        fetching: string;
    };
    /**
     * Text displayed within the button tooltip for each of the possible button statuses
     */
    tooltipText: {
        default: string;
        error: string;
        fetching: string;
    };
    /**
     * url or endpoint which the button will make a GET request against
     */
    url: string;
    /**
     * if passed, the value here will be the max duration a request can take before it will automatically be cancelled.
     * if not passed, requests will not be cancelled
     */
    requestTimeout?: number;
    /**
     * if the button is disabled or not
     */
    disabled?: boolean;
    /**
     *  if you wish to mount the component in an error state
     */
    hasError?: boolean;
    /**
     * if you wish to mount the component in a fetching state
     */
    isFetching?: boolean;
    /**
     * aria label to assign to the button for screen readers
     */
    ariaLabel?: string;
    /**
     * A callback function which will be called on success / failure of the api call
     * {Success} - Will return a response body in success cases
     * {Error} - Will return an error object in error case
     */
    callback: Function;
}

export interface ButtonProps {
    id: string;
    type: string;
    disabled?: boolean;
    status: 'error' | 'fetching' | 'default';
    ariaLabel?: string;
    onClick: Function;
}

export interface ButtonWrapperProps {
    onMouseOver: Function;
    onMouseOut: Function;
}

export interface StatusProps {
    status: string;
    disabled?: boolean;
}
