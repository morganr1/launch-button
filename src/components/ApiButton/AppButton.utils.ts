import { ControlledRequestParams } from './AppButton.types';

export const controlledFetchRequest = async (
    controlRequestParams: ControlledRequestParams
) => {
    const { controllerRef, setRequestStatus, callback, requestTimeout, url } =
        controlRequestParams;
    const controller = new AbortController();
    controllerRef.current = controller;

    try {
        setRequestStatus({ fetching: true, error: false });
        requestTimeout &&
            setTimeout(() => {
                // If timeout provided, we should cancel the in flight request
                // if its duration exceeds the max duration provided via props
                controllerRef.current?.abort();
            }, requestTimeout);
        const response = await fetch(url, {
            signal: controllerRef.current.signal,
        });
        if (response.ok) {
            const body = await response.json();
            callback(body);
            setRequestStatus({
                fetching: false,
                error: false,
            });
            controllerRef.current = null;
        }
    } catch (error) {
        callback({ error });
        setRequestStatus({ fetching: false, error: true });
    }
};
