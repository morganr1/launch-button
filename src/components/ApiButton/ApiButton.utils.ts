import { ControlledRequestParams, ControlledRequest } from './ApiButton.types';

export const controlledFetchRequest: ControlledRequest = async (
    controlRequestParams: ControlledRequestParams
) => {
    const {
        controllerRef,
        setRequestStatus,
        requestCallback,
        requestTimeout,
        url,
    } = controlRequestParams;
    const controller = new AbortController();
    controllerRef.current = controller;

    try {
        setRequestStatus({ fetching: true, error: false });
        requestTimeout &&
            setTimeout(() => {
                controllerRef.current?.abort();
            }, requestTimeout);
        const response = await fetch(url, {
            signal: controllerRef.current.signal,
        });
        if (response.ok) {
            const body = await response.json();
            requestCallback(body);
            setRequestStatus({
                fetching: false,
                error: false,
            });
            controllerRef.current = null;
        }
    } catch (error: unknown) {
        requestCallback({ error });
        setRequestStatus({ fetching: false, error: true });
    }
};
