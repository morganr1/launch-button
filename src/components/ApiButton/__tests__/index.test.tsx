import React from 'react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import { render, screen, act, fireEvent } from '@testing-library/react';
import ApiButton from '../ApiButton';

describe('<ApiButton /> ', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    const clickCallbackMock = jest.fn();
    const buttonProps = {
        disabled: false,
        buttonText: {
            default: 'Launch Rocket',
            error: 'Launch Rocket',
            fetching: 'Launching',
        },
        hasError: false,
        isFetching: false,
        tooltipText: {
            default: 'Ignites the fuel',
            error: 'Ignition error',
            fetching: 'Cancel launch',
        },
        callback: clickCallbackMock,
        url: 'https://httpbin.org/delay/3',
    };

    it('renders with default values', () => {
        const defaultProps = {
            ...buttonProps,
            disabled: undefined,
            hasError: undefined,
            isFetching: undefined,
        };
        const { asFragment } = render(<ApiButton {...defaultProps} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('onClick fire a request against a provided url, displays loading state, subsequent click cancels the request.', async () => {
        global.fetch = jest
            .fn()
            .mockImplementation(() => Promise.resolve({ data: 100 }));
        render(<ApiButton {...buttonProps} />);
        expect(screen.queryByText('Launch Rocket')).toBeInTheDocument();
        userEvent.hover(screen.getByText('Launch Rocket'));
        expect(screen.queryByText('Ignites the fuel')).toBeInTheDocument();
        act(() => {
            fireEvent.click(screen.getByText('Launch Rocket'));
        });
        expect(screen.getByText('Launching')).toBeInTheDocument();
        userEvent.hover(screen.getByText('Launching'));
        expect(screen.queryByText('Cancel launch')).toBeInTheDocument();
        act(() => {
            fireEvent.click(screen.getByText('Launching'));
        });
        expect(screen.getByText('Ignition error')).toBeInTheDocument();
    });

    it('Handles mouse and keyboard interactivity', () => {
        render(<ApiButton {...buttonProps} />);
        const button = screen.getByText('Launch Rocket');
        //Hover
        userEvent.hover(button);
        expect(screen.queryByText('Ignites the fuel')).toBeInTheDocument();
        //Blur
        userEvent.unhover(button);
        expect(screen.queryByText('Ignites the fuel')).not.toBeInTheDocument();
        //Focus
        act(() => {
            fireEvent.focusIn(button);
        });
        expect(screen.queryByText('Ignites the fuel')).toBeInTheDocument();
        //Focus out
        act(() => {
            fireEvent.focusOut(button);
        });
        expect(screen.queryByText('Ignites the fuel')).not.toBeInTheDocument();
        // Dismiss tooltip on `ESCAPE` key press
        userEvent.hover(button);
        expect(screen.queryByText('Ignites the fuel')).toBeInTheDocument();
        act(() => {
            fireEvent.keyDown(button, { key: 'Escape' });
        });
        expect(screen.queryByText('Ignites the fuel')).not.toBeInTheDocument();
        userEvent.hover(button);
        act(() => {
            fireEvent.focus(screen.getByText('Ignites the fuel'));
        });
        expect(screen.getByText('Launch Rocket')).toBeInTheDocument();
        act(() => {
            fireEvent.blur(screen.getByText('Ignites the fuel'));
        });
        expect(screen.getByText('Launch Rocket')).toBeInTheDocument();
    });

    it('initializes in error state if hasError prop is passed as true', () => {
        const errorProps = { ...buttonProps, hasError: true };
        render(<ApiButton {...errorProps} />);
        expect(screen.queryByText('Ignition error')).toBeInTheDocument();
    });

    it('initializes in fetching state if isFetching prop is passed as true', () => {
        const fetchingProps = { ...buttonProps, isFetching: true };
        render(<ApiButton {...fetchingProps} />);
        expect(screen.queryByText('Launching')).toBeInTheDocument();
    });

    it('initalizes in disabled state if disabled prop is passed as true', () => {
        const disabledProps = { ...buttonProps, disabled: true };
        render(<ApiButton {...disabledProps} />);
        expect(screen.queryByText('Launch Rocket')).toBeInTheDocument();
        userEvent.hover(screen.getByText('Launch Rocket'));
        expect(screen.queryByText('Ignites the fuel')).not.toBeInTheDocument();
    });
});
