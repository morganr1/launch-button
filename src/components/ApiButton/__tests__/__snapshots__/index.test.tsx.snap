// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ApiButton />  renders with default values 1`] = `
<DocumentFragment>
  <div
    class="sc-hKMtZM bWwngD"
  >
    <button
      aria-describedby="api-button-tooltip"
      aria-relevant="text"
      class="sc-bczRLJ kzjfbK"
      id="api-button-container"
      type="button"
    >
      <div
        class="sc-jSMfEi ipkHuQ"
      >
        <p
          class="sc-gKXOVf bHFdIC"
        >
          Launch Rocket
        </p>
      </div>
    </button>
  </div>
</DocumentFragment>
`;
